### DISCUSSION TITLE
Remove Sinatra from test stack

### DISCUSSION THREAD

**Discussion by krainboltgreene** [vcr/vcr#867] [Ideas]
We currently use Sinatra to "test" against, but I don't think it's a good idea.

We currently use this setup in at least 50% of all the tests. It requires us to start a server, include Sinatra in our dependencies, and assume Sinatra works fine.

Quite frankly we would be better served just mocking out a request/response cycle.

Points made:
- It's already there and it's not broken.
- Sinatra provides a real HTTP server and real interface testing has benefit.
- We're not using a majority of sinatra, so it's probably far more than needed.
- It can cause deadlocks in certain situations, and threadsaftey is a prominent factor.

**Comment by sigmavirus24** [comment]
Why is it bad?

**Comment by krainboltgreene** [comment]
- It requires us to start a stable Sinatra server (which includes making it work for all versions of ruby we test)
- Sinatra ends up in our dependency list (longer test cycles)
- Bloats our tests.

For the same reason I wouldn't hit Facebook in a rails oauth test and would instead use vcr, I'm not looking to hit a real HTTP server just to test vcr.

**Comment by sigmavirus24** [comment]
Except that stubbing requires you hand-write (or copy-and-paste) test fixtures so you can  properly test what you're no longer testing by hitting a Sinatra server. Maybe Sinatra isn't the most ideal framework to use but it really isn't bloating the tests. I've run them on a 5 year old laptop and didn't have much trouble with them (especially compared to some Rails testsuites I've run).

Further, do you really think having hand-written stub data is a better approach? You wouldn't do that with Facebook for OAuth would you? So why are you settling for it here? To reduce **test** dependencies? To _potentially_ cut-down on test run times? It just seems like an overall bad idea to remove something like a real server from the mix.

**Comment by krainboltgreene** [comment]
Yeah, I think a mocked HTTP server (max 100 lines) is better than running two processes, maintain an app, debugging possible Sinatra issues, and having to manage a custom travis setup :P. Easiest question I'll ever answer!

> You wouldn't do that with Facebook for OAuth would you?

That's exactly what VCR is designed to do, though.

**Comment by krainboltgreene** [comment]
For anyone else watching, please comment. That's why I made an issue first.

**Comment by sigmavirus24** [comment]
> That's exactly what VCR is designed to do, though.

Sort-of. It makes it 100x simpler to maintain them though. As someone who has hand-maintained fixtures like that, it is hell and makes testing absolutely terrible.

> having to manage a custom travis setup

This is kind of a bogus reason and you should know that. Each Travis configuration is "custom" to a degree.

> debugging possible Sinatra issues

Is this really consistently a problem? I haven't seen mention of it over the last several months if so.

> maintain an app

Which does a lot of work for you. Those "max" 100 lines will definitely grow as corner cases you didn't think of crop up because behaviour changes. At the very least to ensure you cover some of the quirks of a server you'll have to replace some tests with multiple tests to catch the corner cases.

**Comment by krainboltgreene** [comment]
Now clearly not all of this is related to Sinatra, but excuse me if I don't think this is somehow normal and want to cut things out.

Edit: I removed a bunch of things I pasted, since they're not relevant.

**Comment by krainboltgreene** [comment]
What this argument comes down to is if Sinatra is worth the weight over a mock of an HTTP request/response cycle when needed.

I'm looking through some of these tests and I'm seeing a lot more unused than used.

**Comment by myronmarston** [comment]
My two cents:

I think it's important that VCR has some integrated tests that use each supported HTTP client to make real HTTP requests against a real HTTP server.  That strategy has served VCR extremely well.  It has exposed many issues that I never would have found by mocking out the request/response cycle.  These tests are also extremely decoupled from the implementation, and have facilitated large refactorings.

All that said, the "boot a new sinatra server in every cucumber scenario" approach has not scaled well, and I think something there needs to change.  The "boot a sinatra server once in the specs and use it for all specs that need to make real requests" approach has worked out very well, though.

> Yeah, I think a mocked HTTP server (max 100 lines)

I'm not quite sure what you mean by this.  Will it be a real HTTP server but be handrolled (rather than using sinatra) to return canned responses?

> is better than running two processes

I don't think using sinatra requires running a separate process.  VCR's "boot sinatra" logic is based on capybara's and originally booted sinatra in a separate thread.  That worked fine until I started testing against Patron, and it would hit deadlocks when the server was running in the same process, so I used a separate process for that.  I think it's reasonable to stop testing patron and switch back to using a thread for the sinatra server, though.

> debugging possible Sinatra issues

I can't think of a time I've had to debug sinatra issues, honestly.

> and having to manage a custom travis setup

What does the use of sinatra have to do with a custom travis setup?  I don't remember anything in our travis scripts that have to do with sinatra.

So, to summarize:
-  I think it would be a bad idea to completely stop testing with real requests, but sinatra certainly isn't the only way to do that.
- The travis build hasn't stayed green because of cucumber timeouts largely due to booting a sinatra server per scenario, and I think something there needs to change.
- As new maintainer, I'm glad @krainboltgreene is taking a step back and reconsidering some of the fundamental things about how VCR was built and tested.

**Comment by sigmavirus24** [comment]
For one thing: double pasting files is clearly cheating =P

For another, I can point out a few other projects that I have worked on or currently work on that have far worse travis setups than that. That looks borderline normal to me.

**Comment by sigmavirus24** [comment]
> As new maintainer, I'm glad @krainboltgreene is taking a step back and reconsidering some of the fundamental things about how VCR was built and tested.

I agree. Fresh eyes are usually good for helping re-evaluate prior decisions. I just personally don't want VCR to lose its stability and maturity as a project

**Comment by myronmarston** [comment]
> Now clearly not all of this is related to Sinatra, but excuse me if I don't think this is somehow normal and want to cut things out.

None of that is related to Sinatra.  It's nearly all stuff that @mislav added in #440 that greatly improves the output of the travis build (e.g. the foldings), ensures all parts of the build run even if a part fails early while still ensuring the build passes or fails properly (so that you always get feedback from the entire build, not just the part of the build up to the first failure) and speeds things up using a cache.

If you want to strip that stuff out of the travis setup, that's reasonable but it has nothing to do with our use of sinatra.

**Comment by myronmarston** [comment]
Also, FWIW, I found that stuff he added useful enough that I ported a bunch of it to RSpec's travis setup.

**Comment by krainboltgreene** [comment]
I assumed that some part of those scripts was devoted to spinning up the webserver. I only recently got all green locally and I've started to look into getting Travis-CI green as well. Point taken.

> I'm not quite sure what you mean by this. Will it be a real HTTP server but be handrolled (rather than using sinatra) to return canned responses?

Just mocking the final method that would actually make a request and having it return a standard HTTP response seems to cover pretty much any of the tests I've looked that that are requiring an entire web application be booted up.

Like, how much of Sinatra is actually being used? By the way, I love Sinatra so this isn't some vendetta against it, and I definitely value hitting a real server...I think. I'm a bit undecided actually regarding that part of testing practice.

Here's what I see locally when I remove sinatra locally and run the tests: 600~ failures due to not being able to make a GET request, 400~ failures due to things I can't at a glance figure out.

**Comment by krainboltgreene** [comment]
I've updated the issue with some of the points made. What I wouldn't give for a machine that did that automatically.

**Comment by myronmarston** [comment]
> Like, how much of Sinatra is actually being used? By the way, I love Sinatra so this isn't some vendetta against it, and I definitely value hitting a real server...I think. I'm a bit undecided actually regarding that part of testing practice.

In the specs, the shared examples that are applied to each supported `hook_into` option and client leverage [this sinatra server](https://github.com/vcr/vcr/blob/master/spec/support/sinatra_app.rb), which is booted at the [bottom of `spec_helper`](https://github.com/vcr/vcr/blob/master/spec/spec_helper.rb#L62) (so you don't have to boot it manually).  The sinatra server has various edge cases that have been built up over the years: responses that have no body (e.g. a 204), a redirect (to facilitate testing that redirects that handled properly), a response with multiple values for the same header, etc.

In the cukes, each scenario that needs a server defines and runs its own server, so that they can function as documentation so the reader can see both the client and server sides.

**Comment by sigmavirus24** [comment]
FWIW

> It can cause deadlocks in certain situations, and threadsaftey is a prominent factor.

Seems vague enough that I disagree with it being summarized that way. It should be summarized as "Patron caused Sinatra to deadlock the tests when run in a separate thread instead of a separate process, but dropping testing for Patron seems reasonable" (Because it Sinatra deadlocking in certain tests has nothing to do with the thread-safety of the library as a whole.)

**Comment by krainboltgreene** [comment]
The features will probably be the thing that keeps Sinatra in.

**Comment by myronmarston** [comment]
> Seems vague enough that I disagree with it being summarized that way. It should be summarized as "Patron caused Sinatra to deadlock the tests when run in a separate thread instead of a separate process, but dropping testing for Patron seems reasonable" (Because it Sinatra deadlocking in certain tests has nothing to do with the thread-safety of the library as a whole.)

To clarify even further: it's not sinatra that was deadlocking, but rather Patron that was deadlocking.  Patron as an HTTP client isn't designed to work with the HTTP server being requested running in the same ruby process. (Which, now that I say that is a pretty odd situation.)  I suspect it has to do with the C bindings in Patron, where it won't yield back to the ruby VM while it's trying to read data off the socket or whatever ... but since it won't yield, the sinatra server can't run to write to the socket.  (That's my best guest as to what happened).

**Comment by myronmarston** [comment]
BTW, @krainboltgreene, I'm in the #vcr channel on irc.freenode.net if you want to chat about any VCR stuff in real time.  I've not been in there much in the last year or so but I'll try to hang out in there while you're getting ramped up on this if you'd find it helpful to have me as a real-time resource whenever you need me.  Or we could do a google chat or hangout or whatever.
